sql/connect/server/src/main/scala/org/apache/spark/sql/connect/ml/MLUtils.scala

```diff
@@ -448,98 +454,149 @@ private[ml] object MLUtils {
+    (
+      classOf[TreeEnsembleModel[_]],
+      Set(
+        "predictLeaf",
+        "trees",
+        "treeWeights",
+        "javaTreeWeights",
+        "getNumTrees",
+        "totalNumNodes",
+        "toDebugString")),
+    (classOf[DecisionTreeClassificationModel], Set("featureImportances")),
+    (classOf[RandomForestClassificationModel], Set("featureImportances", "evaluate")),
+    (classOf[GBTClassificationModel], Set("featureImportances", "evaluateEachIteration")),
+    (classOf[DecisionTreeRegressionModel], Set("featureImportances")),
+    (classOf[RandomForestRegressionModel], Set("featureImportances")),
+    (classOf[GBTRegressionModel], Set("featureImportances", "evaluateEachIteration")),
+
+    // Classification Models
+    (
+      classOf[LogisticRegressionModel],
+      Set("intercept", "coefficients", "interceptVector", "coefficientMatrix", "evaluate")),
+    (classOf[LogisticRegressionSummary], Set("probabilityCol", "featuresCol")),
+    (classOf[BinaryLogisticRegressionSummary], Set("scoreCol")),
+
+    // Regression Models
+    (classOf[LinearRegressionModel], Set("intercept", "coefficients", "scale", "evaluate")),
+    (
+      classOf[LinearRegressionSummary],
+      Set(
+        "predictions",
+        "predictionCol",
+        "labelCol",
+        "featuresCol",
+        "explainedVariance",
+        "meanAbsoluteError",
+        "meanSquaredError",
+        "rootMeanSquaredError",
+        "r2",
+        "r2adj",
+        "residuals",
+        "numInstances",
+        "degreesOfFreedom",
+        "devianceResiduals",
+        "coefficientStandardErrors",
+        "tValues",
+        "pValues")),
+    (classOf[LinearRegressionTrainingSummary], Set("objectiveHistory", "totalIterations")),
+
+    // Clustering Models
+    (classOf[KMeansModel], Set("predict", "numFeatures", "clusterCenters")),
+    (classOf[KMeansSummary], Set("trainingCost")),
+    (
+      classOf[BisectingKMeansModel],
+      Set("predict", "numFeatures", "clusterCenters", "computeCost")),
+    (classOf[BisectingKMeansSummary], Set("trainingCost")),
+
+    // Recommendation Models
+    (
+      classOf[ALSModel],
+      Set(
+        "rank",
+        "itemFactors",
+        "userFactors",
+        "recommendForAllUsers",
+        "recommendForAllItems",
+        "recommendForUserSubset",
+        "recommendForItemSubset")),
+
+    // Association Rules
+    (classOf[FPGrowthModel], Set("associationRules", "freqItemsets")),
+
+    // Feature Models
+    (classOf[StandardScalerModel], Set("mean", "std")),
+    (classOf[MaxAbsScalerModel], Set("maxAbs")),
+    (classOf[MinMaxScalerModel], Set("originalMax", "originalMin")),
+    (classOf[RobustScalerModel], Set("range", "median")))
+
+  private def validate(obj: Any, method: String): Unit = {
+    assert(obj != null)
+    val valid = ALLOWED_ATTRIBUTES.exists { case (cls, methods) =>
+      cls.isInstance(obj) && methods.contains(method)
```

Comment:
So as long as the third-party algorithms follow the same design like extending the base classes, then they can be able to access the ALLOWED_LIST?

Reply:
yes, I think so.
e.g. `XGBoostClassificationModel` extends `ProbabilisticClassificationModel`, so it can access all the allowed methods of `PredictionModel`, `ClassificationModel` and `ProbabilisticClassificationModel`

Reply:
I was about to comment on this. Essentially this requires plugins to provide derived classes for the algorithms they want to override. Which is an ok requirement, but we should probably document this appropriately.